app/vmstorage/servers/vmselect.go

```diff
@@ -249,10 +250,23 @@ func getMaxMetrics(sq *storage.SearchQuery) int {
 	maxMetrics := sq.MaxMetrics
 	maxMetricsLimit := *maxUniqueTimeseries
 	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = 2e9
+		maxMetricsLimit = calculateMaxMetricsLimitByResource(*maxConcurrentRequests, memory.Remaining())
 	}
 	if maxMetrics <= 0 || maxMetrics > maxMetricsLimit {
 		maxMetrics = maxMetricsLimit
 	}
 	return maxMetrics
 }
+
+// calculateMaxMetricsLimitByResource calculate the max metrics limit.
+func calculateMaxMetricsLimitByResource(maxConcurrentRequests, remainingMemory int) int {
+	if maxConcurrentRequests <= 0 {
```

Comment:
afaik, maxConcurrentRequests can't be 0. Otherwise, the rest of the program won't work.